client/components/rewind-credentials-form/index.jsx

```diff
@@ -85,16 +96,36 @@ export class CredentialsForm extends Component {
 
 	handleDelete = () => this.props.deleteCredentials( this.props.siteId, this.props.role );
 
-	togglePrivateKeyField = () =>
-		this.setState( { showPrivateKeyField: ! this.state.showPrivateKeyField } );
+	toggleAdvancedSettings = () =>
+		this.setState( { showAdvancedSettings: ! this.state.showAdvancedSettings } );
+
+	componentWillReceiveProps( nextProps ) {
+		const { rewindState, role, siteSlug } = nextProps;
+		const credentials = find( rewindState.credentials, { role: role } );
+		const nextForm = this.state.form;
+
+		// Populate the fields with data from state if credentials are already saved
+		nextForm.protocol =
+			isEmpty( nextForm.protocol ) && credentials ? credentials.protocol : nextForm.protocol;
+		nextForm.host = isEmpty( nextForm.host ) && credentials ? credentials.host : nextForm.host;
+		nextForm.port = isEmpty( nextForm.port ) && credentials ? credentials.port : nextForm.port;
+		nextForm.user = isEmpty( nextForm.user ) && credentials ? credentials.user : nextForm.user;
+		nextForm.path = isEmpty( nextForm.path ) && credentials ? credentials.path : nextForm.path;
```

Comment:
these are messing with `this.state` which is likely to cause render bugs because React relies on state updates to only occur through `setState`

Reply:
I'm not sure I understand how these are directly manipulating the state. They are manipulating a `nextForm` which is a local var, which then used in `this.setState`. 

Is the issue that I'm feeding it's own value back into it? Or something else?

Can you elaborate?

Reply:
It's because in JavaScript the _value_ of an object is its location in memory, so here's the sequence…

```js
// create a new name binding called `nextForm`
// and set its value to the place in memory where
// the `this.state.form` object is stored
const nextForm = this.state.form

// dereference `nextForm.host`
// this has the same dereference
// as this.state.form.host
nextForm.host = 'blah'

// call setState, but here actually
// this.state.form === nextForm
// so it's possible React may decide
// to skip this update entirely as
// "nothing has changed" yet the
// underlying data has and may render
// inconsistently on further re-renders
this.setState( { form: nextForm } )
```

it's a problem of mutation in JavaScript and it causes a number
of big headaches at times, especially when we mutate a function's
input parameters

```js
const bob = { name: 'bob', age: 33 }
const agedBob = age( bob )

function myData( person ) {
  person.age = person.age + 1
  return person
}

agedBob.age === 34 // good!
bob.age === 34 // uh oh, this should be 33 still!
```

Reply:
by the way, a simple way around this is to move all the setting logic into a new immutable update

```js
const { form } = this.state;

const preferUserInput = ( credential, form ) =>
	isEmpty( form ) && credential
		? credential
		: form

this.setState( mergeWith( { ...credentials }, form, preferUserInput )
```

Reply:
I think https://github.com/Automattic/wp-calypso/pull/21857/files#diff-2b378784868d5dbc05e637afc380ddd6R107 fixes things. It's not quite as elaborate as what we've discussed but I do think it solves the problem. Thoughts?